### DISCUSSION TITLE
HOW TO: min/max query

### DISCUSSION THREAD

**Discussion by BananaMasterz** [simolus3/drift#2137] [Q&A] [answered]
How can I get the max value of a column's table?

Basically this query:

```sql
SELECT MAX(table.table_col) FROM table WHERE table_col_2='something';
```

sorry if this has been asked before but I couldn't find something.

**Comment by simolus3** [comment, marked as answer]
Some aggregate functions are explained [here](https://drift.simonbinder.eu/docs/getting-started/expressions/#aggregate). Your query could be written as follows:

```dart
final max = table.tableCol.max();
final query = selectOnly(table)
  ..addColumns([max])
  ..where(table.tableCol2.equals('something'));
query.map((row) => row.read(max)) /* .getSingle() or .watchSingle() */;
```

**Reply by BananaMasterz** [reply]
is this more optimized / faster than this?

```dart
final query = selectOnly(table)
  ..limit(1)
  ..where(table.tableCol2.equals('something'))
  ..orderBy([OrderingTerm(expression: table.tableCol, mode: OrderingMode.desc)]);
  ```
  does this solution with the orderBy first organize everything and then pulls one? Not sure how either work behind the scenes so I'm curious which would be faster/better

**Reply by simolus3** [reply]
The Dart snippet I've posted corresponds to the SQL query you've written. If you want the full row, I probably would use the limit/order by solution as well (but you need to use `select` instead of `selectOnly` - the purpose of `selectOnly` is to only select selected columns instead of the whole table).

> does this solution with the orderBy first organize everything and then pulls one?

This is hard to answer for sure without looking at which indices and unique or primary keys are on that table. Like all database systems, sqlite3 will try to come up with an efficient program to implement your SQL query. I _think_ sqlite3 is smart enough to not scan the entire row in either case if it has the btrees to avoid that. So it shouldn't make a difference, but I'm not an expert on sqlite3 internals so I can't give a definite answer for that.

**Reply by BananaMasterz** [reply]
I only care about one row and one column. The table has the following columns:

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Kyria/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Kyria/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



id (primary key) | category (string) | amount (int)
-- | -- | --
1 | cat 1 | 1
2 | cat 2 | 23
3 | cat 3 | 12
4 | cat 1 | 4
5 | cat 2 | 2
6 | cat 3 | 7
7 | cat 1 | 5



</body>

</html>


So a very simple table.
I want the max `amount` for the `category='cat 1'` for example in this case would be row 7 and I want a query that would return the `amount` so in this case the query would return `5`

Which solution would you say is best? Hopefully the example I have is not too confusing ...

**Reply by simolus3** [reply]
> Which solution would you say is best?

I would use the one that is easier to read. For me, if I only care about the maximum, using `amount.max()` is more readable than selecting `amount`, ordering by `amount desc` and limiting rows to 1. If you find the other query easier to understand, use that one.

You can start worrying about performance when you're running that query very frequently, or if the table gets huge. In that case, the easiest way to speed up both queries is to add an index on `category`.

**Reply by BananaMasterz** [reply]
Got it... I'll go with your way since it is indeed easier to read. Thanks for your time!
